kernel/src/table_changes/physical_to_logical.rs

```diff
@@ -0,0 +1,191 @@
+
+/// Generates the expression used to convert physical data from the `scan_file` path into logical
+/// data matching the `global_state.logical_schema`
+#[allow(unused)]
+fn get_expression(
+    scan_file: &CdfScanFile,
+    logical_schema: &StructType,
+    all_fields: &[ColumnType],
+) -> DeltaResult<Expression> {
+    let mut generated_columns = get_generated_columns(scan_file)?;
+    let all_fields = all_fields
+        .iter()
+        .map(|field| match field {
+            ColumnType::Partition(field_idx) => {
+                let field = logical_schema.fields.get_index(*field_idx);
+                let Some((_, field)) = field else {
+                    return Err(Error::generic(
+                        "logical schema did not contain expected field, can't transform data",
+                    ));
+                };
+                let name = field.physical_name();
+                let value_expression =
+                    parse_partition_value(scan_file.partition_values.get(name), field.data_type())?;
+                Ok(value_expression.into())
+            }
+            ColumnType::Selected(field_name) => {
+                // Remove to take ownership
+                let generated_column = generated_columns.remove(field_name.as_str());
+                Ok(generated_column.unwrap_or_else(|| ColumnName::new([field_name]).into()))
+            }
+        })
+        .try_collect()?;
+    Ok(Expression::Struct(all_fields))
+}
+
+/// Gets the physical schema that will be used to read data in the `scan_file` path.
+#[allow(unused)]
+fn get_read_schema(scan_file: &CdfScanFile, read_schema: &StructType) -> SchemaRef {
+    if scan_file.scan_type == CdfScanFileType::Cdc {
+        let change_type = StructField::new("_change_type", DataType::STRING, false);
+        let fields = read_schema.fields().cloned().chain(iter::once(change_type));
+        StructType::new(fields).into()
+    } else {
+        read_schema.clone().into()
+    }
+}
+
+/// Reads the data at the `resolved_scan_file` and transforms the data from physical to logical.
+/// The result is a fallible iterator of [`ScanResult`] containing the logical data.
+#[allow(unused)]
+pub(crate) fn read_scan_data(
+    engine: &dyn Engine,
+    resolved_scan_file: ResolvedCdfScanFile,
+    global_state: &GlobalScanState,
+    all_fields: &[ColumnType],
+    predicate: Option<ExpressionRef>,
+) -> DeltaResult<impl Iterator<Item = DeltaResult<ScanResult>>> {
+    let ResolvedCdfScanFile {
+        scan_file,
+        mut selection_vector,
+    } = resolved_scan_file;
+
+    let expression = get_expression(&scan_file, global_state.logical_schema.as_ref(), all_fields)?;
+    let schema = get_read_schema(&scan_file, global_state.read_schema.as_ref());
+    let evaluator = engine.get_expression_handler().get_evaluator(
+        schema.clone(),
+        expression,
+        global_state.logical_schema.clone().into(),
+    );
+
+    let table_root = Url::parse(&global_state.table_root)?;
```

Comment:
aside: feels weird we are parsing here? should global state be a Url instead?

Reply:
throw all these bits into an issue for us to revisit?

Reply:
(but yea serializability mostly makes sense assuming we need to just keep as simple as possible for serialization but we could always just have a Url/String interop: field is Url but trivially serializes as the backing string)